flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java

```diff
@@ -0,0 +1,129 @@
+    private final boolean isExtended;
+
+    public DescribeFunctionOperation(ObjectIdentifier sqlIdentifier, boolean isExtended) {
+        this.sqlIdentifier = sqlIdentifier;
+        this.isExtended = isExtended;
+    }
+
+    public ObjectIdentifier getSqlIdentifier() {
+        return sqlIdentifier;
+    }
+
+    public boolean isExtended() {
+        return isExtended;
+    }
+
+    @Override
+    public String asSummaryString() {
+        Map<String, Object> params = new LinkedHashMap<>();
+        params.put("identifier", sqlIdentifier);
+        params.put("isExtended", isExtended);
+        return OperationUtils.formatWithChildren(
+                "DESCRIBE FUNCTION", params, Collections.emptyList(), Operation::asSummaryString);
+    }
+
+    @Override
+    public TableResultInternal execute(Context ctx) {
+        // DESCRIBE FUNCTION <function> shows all the function properties.
+        Optional<ContextResolvedFunction> functionOpt =
+                ctx.getFunctionCatalog().lookupFunction(UnresolvedIdentifier.of(sqlIdentifier));
+        if (!functionOpt.isPresent()) {
+            throw new ValidationException(
+                    String.format(
+                            "Function with the identifier '%s' doesn't exist.",
+                            sqlIdentifier.asSummaryString()));
+        }
+        ContextResolvedFunction function = functionOpt.get();
+        CatalogFunction catalogFunction = function.getCatalogFunction();
+
+        if (catalogFunction == null && !isExtended) {
+            throw new ValidationException(
+                    String.format(
+                            "Function with the identifier '%s' is a system function which can only be described by using the extended keyword. Use 'DESCRIBE FUNCTION EXTENDED %s' to see its properties.",
+                            sqlIdentifier.asSummaryString(), sqlIdentifier.asSummaryString()));
+        }
+
+        List<List<Object>> rows = new ArrayList<>();
+        if (catalogFunction != null) {
+            rows.add(Arrays.asList("class name", catalogFunction.getClassName()));
+            rows.add(
+                    Arrays.asList(
+                            "function language", catalogFunction.getFunctionLanguage().toString()));
+            rows.add(
+                    Arrays.asList(
+                            "resource uris", catalogFunction.getFunctionResources().toString()));
+            rows.add(Arrays.asList("temporary", function.isTemporary()));
+        }
+
+        if (isExtended) {
+            FunctionDefinition definition = function.getDefinition();
+            rows.add(Arrays.asList("kind", definition.getKind().toString()));
+            rows.add(Arrays.asList("requirements", definition.getRequirements().toString()));
+            rows.add(Arrays.asList("deterministic", definition.isDeterministic()));
+            rows.add(Arrays.asList("constant folding", definition.supportsConstantFolding()));
+            rows.add(
```

Comment:
If we wanted to extend this to include return types, would that just be added to the signature I assume?  I don't think the format of these signatures is well defined, so maybe you're free to do that in the future.

Reply:
I understand the FLIP does not cover that, but why wont we include the Function return type here?

Reply:
@AlanConfluent we could add it to the signature, or add a new row entirely, whatever makes the most sense in the future implementation, but we certainly do have that flexibility 

@pgaref Type inference is a complicated mechanism, and it is not always straightforward on how/what return type to provide for a function. Currently you would need the `CallContext` in order to infer the output type of the result.